### PR TITLE
doc/website/blog: troublshooting migrations

### PR DESCRIPTION
**Pull request by rotemtam**

### CODE REVIEW

**Review by a8m — APPROVED** [review]
🚀 

Well done, Rotem.

File: doc/website/blog/2023-04-10-troubleshooting-migrations.mdx [hunk 1]
```diff
@@ -0,0 +1,275 @@
...
+authors: rotemtam
+tags: [schema, migration, troubleshooting, mttr]
+---
+
+Database schema migrations are an essential part of software development, allowing teams to
+evolve and refine their application's data model over time. However, with schema changes, it's
+not always smooth sailing, and migration failures can be disruptive and challenging to resolve.
+
+In this blog post, we'll explore the common causes of migration failures and demonstrate how
+Atlas can help you recover quickly from such failures and quickly get back on track.
```
Comment (hilakashai) [inline doc/website/blog/2023-04-10-troubleshooting-migrations.mdx:12]:
```suggestion
Atlas can help you quickly recover from such failures and easily get back on track.
```

File: doc/website/blog/2023-04-10-troubleshooting-migrations.mdx [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+
+As much as we'd like to believe that our schema migrations will be executed flawlessly,
```
Comment (hilakashai) [inline doc/website/blog/2023-04-10-troubleshooting-migrations.mdx:14]: I would put this paragraph before the 'In this blogpost' sentence.

File: doc/website/blog/2023-04-10-troubleshooting-migrations.mdx [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+the reality is that things can and do go wrong. Whether it's due to human error, unforeseen
+complications, or technical constraints, migration failures can be a significant source
+of frustration for development teams. Anticipating and preparing for these issues is
+essential to minimize their impact on your project.
+
+### Atlas: Optimized for MTTR
+
+MTTR (mean-time-to-recovery) is a widely accepted metric for measuring the performance
+of teams delivering software. MTTR measures the mean time it takes to restore service
```
Comment (hilakashai) [inline doc/website/blog/2023-04-10-troubleshooting-migrations.mdx:23]:
```suggestion
of teams delivering software. MTTR measures the mean time it takes to restore a service
```

File: doc/website/blog/2023-04-10-troubleshooting-migrations.mdx [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+when a production issue occurs. In the context of schema migrations, this would mean
+measuring how long it takes a team to detect, triage and resolve failures of schema migrations.
+
+Contrary to existing tools, Atlas was designed with failure in mind and comes with some
+useful features to help your team get out of the mud if (and when) a schema migration
+you execute fails. By utilizing these features, your team can greatly reduce MTTR for
```
Comment (hilakashai) [inline doc/website/blog/2023-04-10-troubleshooting-migrations.mdx:29]:
```suggestion
fails. By utilizing these features, your team can greatly reduce MTTR for
```

File: doc/website/blog/2023-04-10-troubleshooting-migrations.mdx [hunk 5]
```diff
@@ -0,0 +1,275 @@
...
+schema change related failures.
+
+### Why do migrations fail?
+
+Let's begin our discussion of troubleshooting schema migration failures by mentioning
+the common causes for migration failures.
+
+1. Syntax errors - a surprisingly common cause for migration failures is syntax errors in the
```
Comment (hilakashai) [inline doc/website/blog/2023-04-10-troubleshooting-migrations.mdx:37]:
```suggestion
1. Syntax errors - A surprisingly common cause for migration failures is syntax errors in the
```
stay consistent 😉

File: doc/website/blog/2023-04-10-troubleshooting-migrations.mdx [hunk 6]
```diff
@@ -0,0 +1,275 @@
...
+
+3. Data-dependent changes - If migrations contain the manipulation of data or the modification
+  of constraints, the operation may fail, depending on existing data in the target database.
+  For example, adding a `NOT NULL` constraint to a column may fail, if that column contains null values:
+  ```
+  mysql> alter table users modify bio varchar(100) not null;
+  ERROR 1138 (22004): Invalid use of NULL value
+  ```
+
+4. Lost connection - in some cases, and depending on the state of the target database
```
Comment (hilakashai) [inline doc/website/blog/2023-04-10-troubleshooting-migrations.mdx:63]:
```suggestion
4. Lost connection - In some cases, and depending on the state of the target database
```

File: doc/website/blog/2023-04-10-troubleshooting-migrations.mdx [hunk 7]
```diff
@@ -0,0 +1,275 @@
...
+
+2. Schema dependent changes - Incorrect assumptions about the current state of the target
+  database can lead to failed migrations when those assumptions are not met.
+  For example, trying to create a table that was already created:
+```
+mysql> create table users (   id int,   name varchar(255) );
+ERROR 1050 (42S01): Table 'users' already exists
+```
+
+3. Data-dependent changes - If migrations contain the manipulation of data or the modification
```
Comment (hilakashai) [inline doc/website/blog/2023-04-10-troubleshooting-migrations.mdx:55]:
```suggestion
3. Data-dependent changes - If migrations manipulate data or modify
```

File: doc/website/blog/2023-04-10-troubleshooting-migrations.mdx [hunk 8]
```diff
@@ -0,0 +1,275 @@
...
+  of constraints, the operation may fail, depending on existing data in the target database.
```
Comment (hilakashai) [inline doc/website/blog/2023-04-10-troubleshooting-migrations.mdx:56]:
```suggestion
constraints, the operation may fail depending on existing data in the target database.
```

File: doc/website/blog/2023-04-10-troubleshooting-migrations.mdx [hunk 9]
```diff
@@ -0,0 +1,275 @@
...
+  For example, adding a `NOT NULL` constraint to a column may fail, if that column contains null values:
```
Comment (hilakashai) [inline doc/website/blog/2023-04-10-troubleshooting-migrations.mdx:57]:
```suggestion
  For example, adding a `NOT NULL` constraint to a column may fail if that column contains null values:
```

File: doc/website/blog/2023-04-10-troubleshooting-migrations.mdx [hunk 10]
```diff
@@ -0,0 +1,275 @@
...
+  No connection. Trying to reconnect...
+  ERROR 2003 (HY000): Can't connect to MySQL server on '0.0.0.0:3306' (61)
+  ERROR:
+  Can't connect to the server
+  ```
+
+### Troubleshooting failures with Atlas
+
+In the next section, we review the capabilities that Atlas provides operators to
+troubleshoot and resolve migration failures. We will discuss:
```
Comment (hilakashai) [inline doc/website/blog/2023-04-10-troubleshooting-migrations.mdx:77]:
```suggestion
troubleshoot and resolve migration failures:
```

File: doc/website/blog/2023-04-10-troubleshooting-migrations.mdx [hunk 11]
```diff
@@ -0,0 +1,275 @@
...
+
+* Status observability - how to understand the current state of the system after a failure.
+* Statement level granularity - how to recover from partial migration failures.
+* Declarative roll-forward - how to use Atlas to automatically create a recovery plan from a failure.
+
+#### Status observability
+
+The first step to solving any failure is being to triage the issue at hand. To assist
```
Comment (hilakashai) [inline doc/website/blog/2023-04-10-troubleshooting-migrations.mdx:85]:
```suggestion
The first step to solving any failure is being able to triage the issue at hand. To assist
```

File: doc/website/blog/2023-04-10-troubleshooting-migrations.mdx [hunk 12]
```diff
@@ -0,0 +1,275 @@
...
+
+* `Migration Status: PENDING` - There are pending migrations.
+* `-- Executed Files:  2 (last one partially)` - the last file was _partially_ applied.
+* The last migration failed with an error: `ERROR: Error 1051 (42S02): Unknown table 'test.non_existing'`
+
+#### Statement-level granularity
+
+As we saw in the example above, in cases where migrations partially fail (only some
+statements succeed) our database schema will be in a *limbo state* of sorts, it's
+neither in the previous nor the next version.  To keep implementations simple, in
```
Comment (hilakashai) [inline doc/website/blog/2023-04-10-troubleshooting-migrations.mdx:135]:
```suggestion
neither in the previous nor the next version. To keep implementations simple, in
```

File: doc/website/blog/2023-04-10-troubleshooting-migrations.mdx [hunk 13]
```diff
@@ -0,0 +1,275 @@
...
+the past, many migration tools have opted to treat migration files as *opaque blobs*,
```
Comment (hilakashai) [inline doc/website/blog/2023-04-10-troubleshooting-migrations.mdx:136]:
```suggestion
the past many migration tools have opted to treat migration files as *opaque blobs*,
```

File: doc/website/blog/2023-04-10-troubleshooting-migrations.mdx [hunk 14]
```diff
@@ -0,0 +1,275 @@
...
+which meaning they cannot provide any assistance in cases of partial failures.
```
Comment (hilakashai) [inline doc/website/blog/2023-04-10-troubleshooting-migrations.mdx:137]:
```suggestion
meaning they cannot provide any assistance in cases of partial failures.
```

File: doc/website/blog/2023-04-10-troubleshooting-migrations.mdx [hunk 15]
```diff
@@ -0,0 +1,275 @@
...
+
+Atlas, on the other hand, actually parses the migration files prior to executing
```
Comment (hilakashai) [inline doc/website/blog/2023-04-10-troubleshooting-migrations.mdx:139]:
```suggestion
Atlas, on the other hand, parses the migration files prior to executing
```

File: doc/website/blog/2023-04-10-troubleshooting-migrations.mdx [hunk 16]
```diff
@@ -0,0 +1,275 @@
...
+  -- 2 sql statements
+```
+
+#### Declarative roll-forward
+
+One of the things people experienced with existing tools immediately notice when
+they start working with Atlas is the absence of *down migrations*. Many migration
+tools expect users to plan a *down migration* parallel to every migration, which
+contains the statements needed to roll back the schema changes for a version.
+This is done, in theory, to allow users to seamlessly return to a previous version
```
Comment (hilakashai) [inline doc/website/blog/2023-04-10-troubleshooting-migrations.mdx:214]:
```suggestion
In theory, this is done to allow users to seamlessly return to a previous version
```

File: doc/website/blog/2023-04-10-troubleshooting-migrations.mdx [hunk 17]
```diff
@@ -0,0 +1,275 @@
...
+
+Let's unpack this command:
+* `url` - is the URL of the target database that we want to modify.
+* `to` - describes the desired state, in this case the migration directory at
+  `file://migrations` at version `targetVersion` - omitting this query parameter
+  will set the desired state at the most recent revision.
+* `dev-url` - Atlas requires a connection to an empty *dev-database,* which it uses to
+  normalize the desired schema. Using the `docker://` URL scheme tells Atlas to spin
+  up and use a fresh Docker container for this purpose.
+* `exclude` - tells Atlas to ignore `atlas_schema_revision` which is a metadata
```
Comment (hilakashai) [inline doc/website/blog/2023-04-10-troubleshooting-migrations.mdx:261]:
```suggestion
* `exclude` - tells Atlas to ignore `atlas_schema_revision` which is a metadata table
```

**Review by hilakashai — COMMENTED** [review]
🚀

File: doc/website/blog/2023-04-10-troubleshooting-migrations.mdx (hunk 3 above)
Comment (rotemtam) [reply doc/website/blog/2023-04-10-troubleshooting-migrations.mdx:23]: this is correct usage , ignoring

**Pull request merged**